### ISSUE TITLE
Change Request: FlatConfig.languageOptions.sourceType should able to be a handler function

### ISSUE THREAD

**Issue by loynoir** [eslint/eslint#16852]
### ESLint version

8.33.0

### What problem do you want to solve?

ESLint use file pattern matching, but file pattern matching is not silver bullet.

So, 
- What about let plugin do some handling, instead of file pattern matching?

For example, file pattern matching can not indicate correct `sourceType` for all scenarios.

https://eslint.org/docs/latest/use/configure/configuration-files-new#configuring-the-javascript-source-type

```js
export default [
    {
        files: ["**/*.js"],
        languageOptions: {
            sourceType: "script"
        }
    }
];
```

### What do you think is the correct solution?

In style of official doc of new config system

https://eslint.org/blog/2022/08/new-config-system-part-2/#custom-parsers-and-parser-options-are-mostly-the-same

```js
import babelParser from "@babel/eslint-parser";

export default [
    {
        files: ["**/*.js", "**/*.mjs"],
        languageOptions: {
            parser: babelParser
        }
    }
];
```

Feat
```js
import {sourceTypeIndicator} from "@some/sourceTypeIndicator";

export default [
    {
        files: ["**/*.js"],
        languageOptions: {
            sourceType: sourceTypeIndicator
        }
    }
];
```

Before feat
```typescript
type FlatConfigLanguageOptionsSourceType = "script" | "module" | "commonjs" | undefined
```

After feat
```typescript
type FlatConfigLanguageOptionsSourceType =
  | 'script'
  | 'module'
  | 'commonjs'
  | undefined
  | ((abosultePath: string) => 'script' | 'module' | 'commonjs' | undefined)
```

Or better
```typescript
type FlatConfigLanguageOptionsSourceType =
  | 'script'
  | 'module'
  | 'commonjs'
  | undefined
  | ((abosultePath: string) => 'script' | 'module' | 'commonjs' | undefined)
  | ((abosultePath: string) => Promise<'script' | 'module' | 'commonjs' | undefined>)
```

### Participation

- [ ] I am willing to submit a pull request for this change.
